### DISCUSSION TITLE
MVC 패턴을 사용하는 이유

### DISCUSSION THREAD

**Discussion by GIVEN53** [2024-woowacourse-study/level-interview#25] [OOP]
1. MVC 패턴의 정의
2. 지켜야 하는 규칙
3. 왜 사용하는지

**Comment by kunsanglee** [comment]
### MVC 패턴의 정의
MVC(Model-View-Controller) 패턴은 소프트웨어 설계에서 사용자 인터페이스를 개발하기 위한 구조적인 패턴 중 하나입니다. 이 패턴은 애플리케이션을 세 가지 주요 구성 요소로 분리하여 개발하는 방법론입니다.

- 모델(Model): 데이터와 비즈니스 로직을 처리하는 부분입니다. 모델은 데이터의 상태를 관리하고, 상태의 변경을 뷰에게 알리며, 필요에 따라 데이터를 변형하는 메소드를 포함합니다.
- 뷰(View): 사용자에게 정보를 표시하는 부분입니다. 뷰는 모델을 기반으로 사용자 인터페이스를 생성하고, 사용자의 입력을 받아 컨트롤러에 전달합니다.
- 컨트롤러(Controller): 사용자의 입력을 처리하고, 모델과 뷰 사이의 상호작용을 관리하는 부분입니다. 컨트롤러는 사용자의 입력에 따라 모델을 업데이트하고, 모델의 상태 변경을 뷰에 반영하도록 지시합니다.

### 지켜야 하는 규칙
- 분리의 원칙: 모델, 뷰, 컨트롤러는 각각 독립적으로 작동해야 하며, 각 구성 요소 간의 의존성은 최소화되어야 합니다.
- 단일 책임 원칙: 각 구성 요소는 자신의 역할에만 집중해야 합니다. 즉, 모델은 데이터 관리, 뷰는 사용자 인터페이스 표시, 컨트롤러는 사용자 입력 처리에만 집중해야 합니다.
- 데이터 흐름: 사용자의 입력은 뷰를 통해 컨트롤러에 전달되고, 컨트롤러는 모델을 업데이트한 후, 모델의 변경 사항은 다시 뷰에 반영되어 사용자에게 표시되어야 합니다.

### MVC 패턴을 적용하면 얻을 수 있는 이점
- 유지보수성: MVC 패턴은 애플리케이션의 구조를 명확하게 분리함으로써, 각 부분을 독립적으로 개발하고 수정할 수 있게 합니다. 이는 유지보수를 용이하게 하고, 개발 과정에서의 협업을 촉진합니다.
- 확장성: 애플리케이션의 각 부분이 분리되어 있기 때문에, 새로운 기능이나 사용자 인터페이스를 추가하기가 더 쉽습니다. 개발자는 기존 코드를 크게 변경하지 않고도 애플리케이션을 확장할 수 있습니다.
- 재사용성: 뷰와 모델이 분리되어 있기 때문에, 동일한 모델에 대해 다양한 뷰를 쉽게 생성할 수 있습니다.

**Comment by seunghye218** [comment]
MVC 패턴은 관심사를 분리하여 코드를 더욱 체계화하고 유지 관리하기 쉽게 만듭니다. 이는 모듈성을 증가시켜 객체의 독립적인 개발 및 테스트를 가능하게 합니다. 이를 통해 테스트 가능성과 확장성이 향상됩니다. 또한 표현과 논리를 명확하게 구분하므로 개발자는 자신의 전문 분야에 집중할 수 있습니다.

**Comment by le2sky** [comment]
MVC 패턴은 유지보수가 편해지는 코드 구성 방식입니다.
요소는 Model, Controller, View가 존재합니다.
Model은 데이터와 관련된 부분이며, View는 사용자에게 보여지는 부분, Controller는 Model과 View를 연결해주는 부분입니다.

MVC 패턴에서는 다음과 같은 규칙을 지켜야합니다.

Model은 Controller와 View에 의존하지 않아야 합니다.
View는 Model에만 의존해야 하며, Controller에는 의존하면 안됩니다.
View가 Model로부터 데이터를 받을 때는, 사용자마다 다르게 보여주어야 하는 데이터에 대해서만 받아야 합니다.
Controller는 Model과 View에 의존해도 괜찮습니다.
View가 Model로부터 데이터를 받을 때, 반드시 Controller에서 받아야 합니다.

그리고, 이러한 MVC 패턴의 목적은 유지보수의 용이성입니다.

**Comment by GIVEN53** [comment]
모델, 뷰, 컨트롤러로 관심사를 분리해서 데이터를 제어하는 패턴

1. Model은 Controller와 View에 의존하지 않아야 한다.
2. View는 Model에만 의존해야 한다.
3. View가 Model로부터 데이터를 받을 때는 사용자마다 다르게 보여줘야 하는 데이터만 받아야 한다.
4. Controller는 Model과 View에 의존해도 된다.
5. View가 Model로부터 데이터를 받을 때 반드시 Controller를 통해서 받아야 한다.

테스트, 확장성, 유지보수 용이
